main/distributed-programming.md

```diff
@@ -42,22 +55,35 @@ lockdown(); // turns the current realm into an immutable realm
 
 JavaScript code runs in the context of a [`Realm`](https://www.ecma-international.org/ecma-262/10.0/index.html#sec-code-realms). A realm is composed of the set of *primordials* (objects and functions of the standard library like `Array.prototype.push`) and of a global object. In a web browser, an iframe is a realm. In Node.js, a Node process is a realm.
 
-For historical reasons, the ECMAScript specification requires the *primordials* to be mutable (`Array.prototype.push = yourFunction` is valid ECMAScript but not recommended). SES provides to turn the current realm into an **immutable realm**, that is a realm within which, the primordials are deeply frozen. It also allows to create **Compartments**.
+For historical reasons, the ECMAScript specification requires the *primordials*
+to be mutable (`Array.prototype.push = yourFunction` is valid ECMAScript but not
+recommended). SES makes it possible to turn the current realm into an
+**immutable realm**, that is a realm within which the primordials are deeply
+frozen. It also allows programs to create **Compartments**.
 
-Compartments are "mini-realms". They have their own dedicated global object and environement, but they inherit the primordials from their parent realm.
+Compartments are "mini-realms". They have their own dedicated global object and environment, but they inherit the primordials from their parent realm.
 
 Agoric "deploy scripts" and smart contract code run in an immutable realm with compartments providing just enough authority to create useful and secure contracts (but not enough authority to do anything unintended or harmful to the participants of the smart contract)
 
 
-
-
 ### harden
 
 `harden` is a function provided by the [`@agoric/harden` package](https://www.npmjs.com/package/@agoric/harden).
 
-In a nutshell, `harden` creates objects that can be trusted. In practice, `harden` performs the deep version of [`Object.freeze`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze) (which only freezes the object's own properties).
-
-Most methods of the Agoric platform require their arguments to be `harden`ed. This is the only way they can provide the security properties they claim to provide.
+In a nutshell, `harden` creates objects that can't be changed. In practice,
+`harden` performs the deep version of
+[`Object.freeze`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze). Freeze
+only freezes the object's own properties, so harden is necessary in order to
+ensure that no further changes can be made. It is useful in distributed systems
+to distinguish objects with mutable state from objects that are
+unchanging. Objects that can change over time are extremely useful, but in order
```

Comment:
The rest of this paragraph was a little confusing to me because of the vagueness.

Reply:
In a lot more explicit detail, what I was trying to say was this:

Why is a deep freeze operation useful?

* Objects with mutable state don't migrate. 
* If you want to interact with such an object, you have to send data to it. 
* The data you're sending can't itself consist of mutable objects.
* You can send primitive values directly.
* If we want mobile data to have more structure, we can make records, but those must be transitively frozen in order to be passed around.

Does that make more sense? Is it worth saying?

Reply:
I dropped most of this.